File: py/examples/counter_multicast.py
# Mode / Multicast
# Launch the server in #multicast #mode to synchronize browser state across a user's clients.
# Open `/demo` in multiple browsers and watch them synchronize in realtime.
# ---
from h2o_wave import main, app, Q, ui


@app('/demo', mode='multicast')
async def serve(q: Q):
    if not q.client.initialized:
        if q.user.count is None:
            q.user.count = 0

        q.page['example'] = ui.form_card(box='1 1 2 1', items=[
            ui.button(name='increment', label=f'Count={q.user.count}')
        ])
        q.client.initialized = True

    if q.args.increment:
        q.user.count += 1
        q.page['example'].increment.label = f'Count={q.user.count}'

    await q.page.save()

File: r/R/examples/counter_multicast.R
#library(h2owave)

app("/demo",mode='multicast')
serve <- function(qo)
{
    qo$user$test <- ifelse(is.null(qo$user$test),0, qo$user$test)
    ifelse(qo$args$increment == TRUE,qo$user$test <- qo$user$test + 1,qo$user$test <- 0)
    if(length(qo$args) == 0 || qo$args$increment == FALSE){
        qo$page$add_card("example",ui_form_card(
                                                box='1 1 12 10'
                                                ,items = list(
                                                              ui_button(
                                                                        name='increment'
                                                                        ,label=paste0('Count= ',qo$user$test)
                                                                        ,primary = TRUE
                                                              )
                                                )
                                                ))
    }
    else {
        qo$page$set("example","items","0","button","label",paste0("Count=",qo$user$test))
    }
    qo$page$save()
}
